在KaimingNormal/Uniform中增加mode参数

Qin-sx opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->

User Experience

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] -->

Improvements

### Description
<!-- Describe what you’ve done -->

**当前问题**  
paddle.nn.initializer.KaimingNormal目前只fan_in模式，不支持fan_out模式，fan_out需要自己计算out_feature值后填入fan_in参数  

**解决方案**  
paddle.nn.initializer.KaimingNormal中增加一个mode参数：
- mode参数可以设置为'fan_in'或者'fan_out'
- 在mode='fan_in'时，fan_in参数才会生效
- 在mode='fan_out'时，会自动计算out feature值用于计算

Qin-sx committed:
added fan_out para in kaiming normal init

	modified:   python/paddle/nn/initializer/kaiming.py
	modified:   test/legacy_test/test_initializer.py

Qin-sx commented:
docs: https://github.com/PaddlePaddle/docs/pull/7324  
PaConver: https://github.com/PaddlePaddle/PaConvert/pull/594

zhwesky2010 commented on python/paddle/nn/initializer/kaiming.py:
```diff
@@ -85,6 +87,17 @@ def __init__(
+            )
+        if self._mode == 'fan_out' and self._fan_in is not None:
+            raise ValueError(
+                "The mode of MSRAInitializer is 'fan_out', "
```
KaimingNormal/KaimingUniform

Qin-sx replied:
嗯，收到，已修改

Qin-sx committed:
updated func name in ValueError

	modified:   python/paddle/nn/initializer/kaiming.py

Qin-sx committed:
added mode para in KaimingUniform for fan_out

	modified:   python/paddle/nn/initializer/kaiming.py
	modified:   test/legacy_test/test_initializer.py

zhwesky2010 commented on test/legacy_test/test_initializer.py:
```diff
@@ -1671,6 +1671,68 @@ def test_msra_initializer(self, dtype="float32"):
+        tensor = paddle.zeros([16, 1024])
+        tensor.stop_gradient = False
+
+        msra_ = paddle.nn.initializer.KaimingNormal(mode='fan_out')
```
增加一个mode='fan_in'的case，测试充分

zhwesky2010 commented:
@Qin-sx CI可能随机挂了，注意rerun下

Qin-sx commented:
> @Qin-sx CI可能随机挂了，注意rerun下

嗯，收到，已重跑

zhwesky2010 commented on python/paddle/nn/initializer/kaiming.py:
```diff
@@ -62,6 +62,7 @@ class MSRAInitializer(Initializer):
         seed (int32, optional): random seed. Default is 0.
         negative_slope (float, optional): negative_slope (only used with leaky_relu). Default is 0.0.
         nonlinearity(str, optional): the non-linear function. Default is relu.
+        mode(str, optional): the mode of initialization, can be 'fan_in' or 'fan_out'. When set to 'fan_in', the fan_in parameter is used for initialization. When set to 'fan_out', the out_features of trainable Tensor will be used. Default is 'fan_in'.
```
这个逻辑是不是得这么改一下，如果设置了 fan_in，则fan_in生效，mode不生效。如果没设置 fan_in，则mode生效。

fan_in与fan_out不可同时设置，给出明确的报错提示。

Qin-sx replied:
嗯，我想的逻辑是以mode参数为判断标准，mode='fan_in'时，fan_in参数生效。mode='fan_out'时，fan_in参数不生效。感觉以fan_in参数为判断标准，逻辑不太好解释。
```python
        if self._mode == 'fan_in':
            fan_in = f_in if self._fan_in is None else self._fan_in
        if self._mode == 'fan_out':
            fan_in = f_out
```
以mode参数为判断标准的报错逻辑在这里
```python
        if self._mode not in ['fan_in', 'fan_out']:
            raise ValueError(
                "The mode of KaimingNormal/KaimingUniform should be 'fan_in' or 'fan_out', "
                f"but received {self._mode}."
            )
        if self._mode == 'fan_out' and self._fan_in is not None:
            raise ValueError(
                "The mode of KaimingNormal/KaimingUniform is 'fan_out', "
                "but fan_in is set. Please set fan_in to None."
            )
```

zhwesky2010 replied:
OK，这样更合理

zhwesky2010 reviewed (approved):
LGTM
